libs/ledger/tests/consensus/pos_tests.cpp

```diff
@@ -0,0 +1,258 @@
+    ret->previous_hash              = genesis->hash;
+    ret->miner_id                   = mining_identity_;
+    ret->timestamp = GetTime(fetch::moment::GetClock("default", fetch::moment::ClockType::SYSTEM));
+
+    // Need to 'trick' the entropy as appearing as an aeon beginning
+    if (desired_block_number == 1)
+    {
+      ret->block_entropy.confirmations[ConstByteArray("test")];
+      assert(!ret->block_entropy.confirmations.empty());
+    }
+    else
+    {
+      // This relies on generating blocks in order
+      ret->previous_hash = chain_.GetHeaviestBlock()->hash;
+    }
+
+    for (auto const &i : qual_)
+    {
+      ret->block_entropy.qualified.insert(i.identifier());
+    }
+
+    ret->weight = consensus_->GetBlockGenerationWeight(*ret, chain::Address{mining_identity_});
+
+    ret->UpdateDigest();
+    ret->miner_signature = cabinet_priv_keys_[0]->Sign(ret->hash);
+
+    chain_.AddBlock(*ret);
+
+    return ret;
+  }
+
+  fetch::crypto::mcl::details::MCLInitialiser init_before_others_{};
+  ConsensusPtr                                consensus_;
+  Signers                                     cabinet_priv_keys_;
+  Members                                     cabinet_;  // The identities that are in the cabinet
+  Members                                     qual_;     // The identities that are in qual
+  StakeManagerPtr                             stake_;
+  BeaconSetupServicePtr                       beacon_setup_;
+  BeaconServicePtr                            beacon_;
+  MainChain                                   chain_;
+  StorageInterface                            storage_;
+  Identity                                    mining_identity_;
+  uint64_t const                              aeon_period_{10};
+  uint64_t const                              max_cabinet_size_{10};
+  uint64_t const                              block_interval_ms_{5000};
+  NotarisationPtr                             notarisation_;
+};
+
+TEST_F(ConsensusTests, test_valid_block)
+{
+  BlockPtr block = ValidNthBlock(1);
+
+  ASSERT_EQ(consensus_->ValidBlock(*block), ledger::ConsensusInterface::Status::YES);
+}
+
+TEST_F(ConsensusTests, test_loose_block)
+{
+  BlockPtr block = ValidNthBlock(1);
+  block->block_number++;
+
+  ASSERT_EQ(consensus_->ValidBlock(*block), ledger::ConsensusInterface::Status::NO);
+}
+
+TEST_F(ConsensusTests, test_entropy_block_number_mismatch)
+{
+  BlockPtr block = ValidNthBlock(1);
+  block->block_entropy.block_number++;
+
+  ASSERT_EQ(consensus_->ValidBlock(*block), ledger::ConsensusInterface::Status::NO);
+}
+
+TEST_F(ConsensusTests, test_hash_empty)
+{
+  BlockPtr block = ValidNthBlock(1);
+  block->hash    = "";
+
+  ASSERT_EQ(consensus_->ValidBlock(*block), ledger::ConsensusInterface::Status::NO);
+}
+
+TEST_F(ConsensusTests, test_wrong_miner_identity)
+{
+  BlockPtr block  = ValidNthBlock(1);
+  block->miner_id = qual_[1];
+
+  ASSERT_EQ(consensus_->ValidBlock(*block), ledger::ConsensusInterface::Status::NO);
```

Comment:
Is this to test incorrect signature? Because I think it also fails in invalid weight.

Reply:
Ahh, you're right. Good catch